{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "#declare path to test genomes\n",
    "path='C:\\\\Users\\\\Matt\\\\Desktop\\\\Bioinfo Project\\\\testGenomes\\\\'\n",
    "#put all test genome flat file names into an array\n",
    "fastafiles=[]\n",
    "for filename in os.listdir(path):\n",
    "    if filename[-6:]==\".fasta\":\n",
    "        fastafiles.append(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading in Matricies\n"
     ]
    }
   ],
   "source": [
    "#load scoring matrices\n",
    "print(\"Loading in Matricies\")\n",
    "loadDataPath='C:\\\\Users\\\\Matt\\\\Desktop\\\\Bioinfo Project\\\\loadData\\\\'\n",
    "orf1matrix=pd.read_pickle(loadDataPath+\"orf1matrix.pckl\")\n",
    "allORFmatrix=pd.read_pickle(loadDataPath+\"allORFmatrix.pckl\")\n",
    "with open(loadDataPath+\"thresholdScores.txt\") as f:\n",
    "    orf1ScoreThreshold=float(f.readline())\n",
    "    allORFScoreThreshold=float(f.readline())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def isolateGenomeToRNA(filename):\n",
    "    with open(filename,'r') as f:\n",
    "        fileString=f.read()\n",
    "        name=fileString.split(\"\\n\")[0]\n",
    "        fileString=fileString[len(name):].replace(\"\\n\",\"\").lower()\n",
    "    #ISOLATE GENOME\n",
    "    genome=re.sub(r'[^a-z]+',\"\",fileString)\n",
    "    \n",
    "    #CHANGE T BASE TO U TO REPRESENT AS RNA\n",
    "    genome=genome.replace('t','u')\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scoreSequence(seq,matrix):\n",
    "    positions=[0,1,2,3,4,5,6,7,8,9,13,14,15,16,17,18,19,20,21,22,23,24]\n",
    "    seq=seq.upper()\n",
    "    score=1\n",
    "    for i in positions:\n",
    "        try:\n",
    "            positionScore=matrix[seq[i]][i]\n",
    "        except:\n",
    "            positionScore=0\n",
    "        if positionScore==0:\n",
    "            score*=.000001\n",
    "        else:\n",
    "            score*=positionScore\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "#FUNCTION FOR FINDING THE END OF AN ORF BY LOOKING FOR THE FIRST SOPT CODON IN THE READING FRME\n",
    "def findNextStop(seq,start,end):\n",
    "        stopCodons=[\"uag\",\"uga\",\"uaa\"]\n",
    "        for i in range(start,end,3):\n",
    "            if seq[i:i+3] in stopCodons:\n",
    "                return i+3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate(rawGenome):\n",
    "    #dictionary to store open reading frames\n",
    "    openReadingFrames={}\n",
    "    \n",
    "    #Locate ORF1a start\n",
    "    for i in range(len(rawGenome)):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            try:\n",
    "                seq=rawGenome[i-10:i+15]\n",
    "            except:\n",
    "                start1a=i\n",
    "                break\n",
    "            score=scoreSequence(seq,orf1matrix)\n",
    "            if score>orf1ScoreThreshold:\n",
    "                start1a=i\n",
    "    \n",
    "    #Locate ends of ORF1a/b and add orf1a/b to dictionary\n",
    "    end1a=findNextStop(rawGenome,start1a,len(rawGenome))\n",
    "    openReadingFrames[\"1a\"]=str(start1a+1) + \",\" + str(end1a)\n",
    "    end1b=findNextStop(rawGenome,end1a-1,len(rawGenome))\n",
    "    openReadingFrames[\"1b\"]=str(start1a+1) + \",\" + str(end1b)\n",
    "    \n",
    "    #Loacte ORF2a start\n",
    "    for i in range(end1b,len(rawGenome)):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            seq=rawGenome[i-10:i+15]\n",
    "            score=scoreSequence(seq, allORFmatrix)\n",
    "            if score>allORFScoreThreshold:\n",
    "                start2a=i+1\n",
    "                break\n",
    "    #Locate ORF2a end\n",
    "    end2a=findNextStop(rawGenome,start2a-1,len(rawGenome))\n",
    "    #add ORF2a to dictionary\n",
    "    openReadingFrames[\"2a\"]=str(start2a)+\",\"+str(end2a)\n",
    "    \n",
    "    #Locate ORF2b start\n",
    "    for i in range(start2a+1,len(rawGenome)):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            seq=rawGenome[i-10:i+15]\n",
    "            score=scoreSequence(seq, allORFmatrix)\n",
    "            if score>allORFScoreThreshold:\n",
    "                start2b=i+1\n",
    "                break\n",
    "    #Locate ORF2b end\n",
    "    end2b=findNextStop(rawGenome,start2b-1,len(rawGenome))\n",
    "    #add ORF2b to dictionary\n",
    "    openReadingFrames[\"2b\"]=str(start2b)+\",\"+str(end2b)\n",
    "    \n",
    "    #Locate ORF3 start\n",
    "    for i in range(end2a-190,end2a):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            seq=rawGenome[i-10:i+15]\n",
    "            score=scoreSequence(seq,allORFmatrix)\n",
    "            if score>allORFScoreThreshold:\n",
    "                start3=i+1\n",
    "                break\n",
    "    #Locate ORF3 end\n",
    "    end3=findNextStop(rawGenome,start3-1,len(rawGenome))\n",
    "    #Add ORF3 to dictionary\n",
    "    openReadingFrames[\"3\"]=str(start3)+\",\"+str(end3)\n",
    "    \n",
    "    #Locate ORF4 start\n",
    "    start4=-1\n",
    "    trace=0\n",
    "    while start4==-1:\n",
    "        for i in range(end3-(230+trace),end3-100):\n",
    "            if rawGenome[i:i+3]=='aug':\n",
    "                seq=rawGenome[i-10:i+15]\n",
    "                score=scoreSequence(seq,allORFmatrix)\n",
    "                if score>allORFScoreThreshold:\n",
    "                    start4=i+1\n",
    "                    break\n",
    "        trace+=10\n",
    "    #Locate ORF4 end\n",
    "    end4=findNextStop(rawGenome,start4-1,len(rawGenome))\n",
    "    #Add ORF4 to dictionary\n",
    "    openReadingFrames[\"4\"]=str(start4)+\",\"+str(end4)\n",
    "    \n",
    "    #Locate ORF5a start\n",
    "    trace=0\n",
    "    start5=-1\n",
    "    while start5==-1:\n",
    "        for i in range(end4+(3+trace),end4+60):\n",
    "            if rawGenome[i:i+3]=='aug':\n",
    "                seq=rawGenome[i-10:i+15]\n",
    "                score=scoreSequence(seq,allORFmatrix)\n",
    "                if score>allORFScoreThreshold:\n",
    "                    start5=i+1\n",
    "                    break\n",
    "        if trace==0: trace=-13\n",
    "        elif trace==-13: trace=50\n",
    "        else: break\n",
    "    #Locate ORF5a end\n",
    "    end5=findNextStop(rawGenome,start5-1,len(rawGenome))\n",
    "    #Add ORF5a to dictionary\n",
    "    openReadingFrames[\"5\"]=str(start5)+\",\"+str(end5)\n",
    "    \n",
    "    #Locate ORF6 start\n",
    "    for i in range(end5-30,len(rawGenome)):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            seq=rawGenome[i-10:i+15]\n",
    "            score=scoreSequence(seq,allORFmatrix)\n",
    "            if score>allORFScoreThreshold:\n",
    "                start6=i+1\n",
    "                break\n",
    "    #Locate ORF6 end\n",
    "    end6=findNextStop(rawGenome,start6-1,len(rawGenome))\n",
    "    #Add ORF6 to dictionary\n",
    "    openReadingFrames[\"6\"]=str(start6)+\",\"+str(end6)\n",
    "    \n",
    "    #Locate ORF7 start\n",
    "    for i in range(end6-30,len(rawGenome)):\n",
    "        if rawGenome[i:i+3]=='aug':\n",
    "            seq=rawGenome[i-10:i+15]\n",
    "            score=scoreSequence(seq,allORFmatrix)\n",
    "            if score>allORFScoreThreshold:\n",
    "                start7=i+1\n",
    "                break\n",
    "    #Locate ORF7 end\n",
    "    end7=findNextStop(rawGenome,start7-1,len(rawGenome))\n",
    "    #Add ORF7 to dictionary\n",
    "    openReadingFrames[\"7\"]=str(start7)+\",\"+str(end7)\n",
    "    \n",
    "    return openReadingFrames\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotating Genomes...\n",
      "KF287133.fasta\n",
      "1a: 166,7785\n",
      "1b: 166,12155\n",
      "2a: 12157,12927\n",
      "2b: 12162,12383\n",
      "3: 12780,13544\n",
      "4: 13325,13861\n",
      "5: 13872,14474\n",
      "6: 14459,14983\n",
      "7: 14973,15344\n",
      "\n",
      "KY373214.fasta\n",
      "1a: 192,7310\n",
      "1b: 192,11680\n",
      "2a: 11682,12452\n",
      "2b: 11687,11908\n",
      "3: 12305,13069\n",
      "4: 12850,13386\n",
      "5: 13397,13996\n",
      "6: 13981,14505\n",
      "7: 14495,14866\n",
      "\n",
      "MF326985.fasta\n",
      "1a: 189,7400\n",
      "1b: 189,11770\n",
      "2a: 11772,12542\n",
      "2b: 11777,11998\n",
      "3: 12395,13159\n",
      "4: 12940,13476\n",
      "5: 13487,14089\n",
      "6: 14074,14598\n",
      "7: 14588,14959\n",
      "\n",
      "MG011718.fasta\n",
      "1a: 189,7307\n",
      "1b: 189,11677\n",
      "2a: 11679,12449\n",
      "2b: 11684,11905\n",
      "3: 12302,13066\n",
      "4: 12847,13383\n",
      "5: 13394,13996\n",
      "6: 13981,14505\n",
      "7: 14495,14866\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotating Genomes...\")\n",
    "for i in range(len(fastafiles)):\n",
    "    print(fastafiles[i])\n",
    "    annotatedGenome=annotate(isolateGenomeToRNA(path+fastafiles[i]))\n",
    "    for key in annotatedGenome.keys():\n",
    "        print(key + \": \" + annotatedGenome[key])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done...\n"
     ]
    }
   ],
   "source": [
    "print(\"Done...\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
