{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports and reading in files + path setup\n",
    "import os\n",
    "import pandas as pd\n",
    "flatfiles=[]\n",
    "for filename in os.listdir(\"genomes\"):\n",
    "    flatfiles.append(filename)\n",
    "\n",
    "path='C:\\\\Users\\\\Matt\\\\Desktop\\\\Bioinfo Project\\\\genomes\\\\'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parseFile2(filename,path): \n",
    "    contents=open(path+filename).read()\n",
    "    #Opens fasta file\n",
    "#    path = \"C:\\\\Users\\\\camos333\\\\Desktop\\\\AF046869.txt\"\n",
    "#    f=open(path, \"r\")\n",
    "#    if f.mode == \"r\":\n",
    "#        contents=f.read()\n",
    "    \n",
    "    #Imports Regular Expressions\n",
    "    import re\n",
    "    #import pandas as pd\\\n",
    "    \n",
    "    #Finds the character number that the word ORIGIN ends and the nucleic acid sequence begins\n",
    "    def nucleostartno(dna):\n",
    "        startno = re.search(r'\\nORIGIN', dna).end()\n",
    "        return startno\n",
    "    \n",
    "    #Trims the sequence to remove spaces, numbers, slashes, and replaces t to u respectively\n",
    "    def justactg1234ss(dna):\n",
    "        nucleo = re.sub(r'([\\s])','',dna)\n",
    "        return nucleo\n",
    "    def justactgss(dna):\n",
    "        nucleo = re.sub(r'\\d','',dna)\n",
    "        return nucleo\n",
    "    def justactg(dna):\n",
    "        nucleo = re.sub(r'/','',dna)\n",
    "        return nucleo\n",
    "    def justacug(dna):\n",
    "        nucleo = re.sub(r't','u',dna)\n",
    "        return nucleo\n",
    "    #Sets contents to trim itself to everything past the word ORIGIN\n",
    "    cutoffORIGIN = str(contents[((nucleostartno(contents))+7):])\n",
    "    \n",
    "    #Goes through and uses those 4 functions\n",
    "    actgstring1 = justactg1234ss(cutoffORIGIN)\n",
    "    actgstring2 = justactgss(actgstring1)\n",
    "    actgstring3 = justactg(actgstring2)\n",
    "    actgstring = str.upper(justacug(actgstring3))\n",
    "    #quick capitalization\n",
    "    \n",
    "    #actgstring is the isolated nucleotide sequence from the GenBank file.\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    '''FROME HERE ON IS WHERE IT ISOLATES FROM actgstring WHERE EACH CDS IS'''\n",
    "    \n",
    "    #Function that splits Genbank File by where the word 'CDS' is\n",
    "    def CDSsplit1(dna):\n",
    "        separate = re.split(r'CDS', dna)\n",
    "        return separate\n",
    "    \n",
    "    x = (CDSsplit1(contents))\n",
    "    \n",
    "    #Cuts off everything in the GenBank file thats before the first 'CDS'\n",
    "    x.pop(0)\n",
    "    \n",
    "    #Once CDS is found, these will select and trim the range to produce the starting number, ensuring there are not carrots.\n",
    "    def CDSstartnucleo1(dna):\n",
    "        startno = ''.join(re.findall(r'\\s\\d*[.]{2}',dna))\n",
    "        return startno\n",
    "    def CDSstartnucleo2(dna):\n",
    "        startno = re.sub(r'[.]','',dna)\n",
    "        return startno\n",
    "    \n",
    "    #Once CDS is found, these will select and trim the range to produce the ending number\n",
    "    def CDSendnucleo1(dna):\n",
    "        startno = ''.join(re.findall(r'[.]{2}\\d*',dna))\n",
    "        return startno\n",
    "    def CDSendnucleo2(dna):\n",
    "        startno = re.sub(r'[.]','',dna)\n",
    "        return startno\n",
    "    \n",
    "    def augfinder(dna):\n",
    "        startno = re.search(r'AUG',dna).start()\n",
    "        return startno\n",
    "    '''EDIT NEXT LINE FOR MASS GENERIC FILE MANIPULATION'''\n",
    "    #Creates the file (if it does not exist) 'CleanedCDS.txt'. I don't know how you'd want to do the file manipulation, so in order for you to execute this file in mass you might want to change how the name thing works\n",
    "\n",
    "    \n",
    "    #Prelimenary lists\n",
    "    truestartlist = []\n",
    "    falsestartlist = []\n",
    "    \n",
    "    #Big Boi lists\n",
    "    trueStarts = []\n",
    "    falseStarts = []\n",
    "    \n",
    "    #This for loop takes the split GenBank file and isolates the correct sequences from the main 'actgstring' string using the numbers from the CDS start and end numbers\n",
    "    for i in x:\n",
    "    #Trims 13 characters up from 'CDS', which is exactly when the numbers start that reference where they are in the genome. 30 characters up is where the numbers have certainly ended. It's fine if the numbers are not exact\n",
    "        i = i[10:30]\n",
    "    #Ensures the those carrot CDS's are not included\n",
    "        if re.search(r'\\s\\d*[.]{2}',i) is not None:\n",
    "    #Isolates start number and end number using the functions created earlier\n",
    "            startno = int(CDSstartnucleo2(CDSstartnucleo1(i)))\n",
    "    #            endno = int(CDSendnucleo2(CDSendnucleo1(i)))\n",
    "    #References 'actgstring' and trims it to only include the coding range for it's respective CDS.\n",
    "            seqt = actgstring[(startno - 10):(startno + 8)]\n",
    "            truestartno = startno-1\n",
    "            truestartlist.append(truestartno)\n",
    "    #LAYOUT. IF YOU THINK ANOTHER LAYOUT WOULD BE MORE USEFUL, THIS IS WHERE YOU NEED TO CHANGE IT. Prepares the layout of text_file. 'entry1' includes 'CDS ###..###' and 'entry2' (after a newline \\n) holds the actual coding region. I made 3 newlines in between each CDS.\n",
    "            trueStarts.append(seqt)\n",
    "    #puts true starts into the correct spot in the dictionary startsoutput\n",
    "            \n",
    "    \n",
    "    \n",
    "    #for use in loop below\n",
    "    location = -1\n",
    "    \n",
    "    # Loop while true, finds AUGs in raw actg sequence and checks if they are in the true list, if so, they are deleted.\n",
    "    while True:\n",
    "        # Advance location by 1.\n",
    "        location = actgstring.find(\"AUG\", location + 1)\n",
    "        \n",
    "        if location == -1: break\n",
    "    \n",
    "    \n",
    "        if not location in truestartlist: falsestartlist.append(location)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #constructs falseStarts\n",
    "    for n in falsestartlist:\n",
    "        falseStarts.append(actgstring[(n - 9):(n + 9)])\n",
    "    \n",
    "    \n",
    "    \n",
    "    #constructs final dictionary\n",
    "    Final = {\"trueStarts\":trueStarts,\"falseStarts\":falseStarts}\n",
    "    #print (Final)\n",
    "    return Final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parseFile(filename, path):\n",
    "    fileString=open(path+filename).read()\n",
    "    \n",
    "    #AQUIRE START INDICIES\n",
    "    startIndex=fileString.find(\"CDS\")\n",
    "    starts=[]\n",
    "    while startIndex>0:\n",
    "        temp=fileString[startIndex+16:fileString.find(\".\",startIndex)]\n",
    "        if temp[0]!='>' and temp[0]!='<':\n",
    "            starts.append(int(temp))\n",
    "            #print(filename)\n",
    "        startIndex=fileString.find(\"CDS\",startIndex+1)\n",
    "    #print(starts)\n",
    "    \n",
    "    #ISOLATE GENOME\n",
    "    genome=\"\"\n",
    "    for char in fileString[fileString.find(\"ORIGIN\"):len(fileString)]:\n",
    "        if char=='t'or char=='g' or char=='a' or char=='c':\n",
    "            genome+=char\n",
    "    \n",
    "    #CHANGE T BASE TO U TO REPRESENT AS RNA\n",
    "    genomeRNA=\"\"\n",
    "    for char in genome:\n",
    "        if char=='t':\n",
    "            genomeRNA+='u'\n",
    "        else:\n",
    "            genomeRNA+=char\n",
    "    \n",
    "    #Get all start codons in genome\n",
    "    startCodonIndicies=[]\n",
    "    for i in range(10,len(genomeRNA)-10):\n",
    "        #print(genome[i:i+3])\n",
    "        if genomeRNA[i:i+3]=='aug':\n",
    "            startCodonIndicies.append(i+1)\n",
    "    \n",
    "    falsePositiveStarts=[]\n",
    "    for index in startCodonIndicies:\n",
    "        if index in starts:\n",
    "            pass\n",
    "        else:\n",
    "            falsePositiveStarts.append(index)\n",
    "    \n",
    "    trueStarts=[]\n",
    "    for index in starts:\n",
    "        trueStarts.append(genomeRNA[index-10:index+8]+filename)\n",
    "    falseStarts=[]\n",
    "    for index in falsePositiveStarts:\n",
    "        falseStarts.append(genomeRNA[index-10:index+8]+filename)\n",
    "    \n",
    "    return {\"trueStarts\": trueStarts, \"falseStarts\": falseStarts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True start codon identities by position\n",
      "    0    1    2    3    4    5    6    7    8    9    10   11   12   13   14  \\\n",
      "A   98    0  204  107   34  487  420  203  403  717    0    0  100  102  208   \n",
      "C  122  204  108  104  280    9    0  305  100    0    0    0   99  371   22   \n",
      "G    9  307   99  113  115   10  297  200    9    0    0  717  329  128  189   \n",
      "U  488  206  306  393  288  211    0    9  205    0  717    0  189  116  298   \n",
      "\n",
      "    15   16   17  \n",
      "A  185  207    8  \n",
      "C    8  210   27  \n",
      "G  284  299  468  \n",
      "U  240    1  214  \n",
      "\n",
      "True start codon frequency by position\n",
      "     0     1     2     3     4     5     6     7     8    9    10   11    12  \\\n",
      "A  0.14  0.00  0.28  0.15  0.05  0.68  0.59  0.28  0.56  1.0  0.0  0.0  0.14   \n",
      "C  0.17  0.28  0.15  0.15  0.39  0.01  0.00  0.43  0.14  0.0  0.0  0.0  0.14   \n",
      "G  0.01  0.43  0.14  0.16  0.16  0.01  0.41  0.28  0.01  0.0  0.0  1.0  0.46   \n",
      "U  0.68  0.29  0.43  0.55  0.40  0.29  0.00  0.01  0.29  0.0  1.0  0.0  0.26   \n",
      "\n",
      "     13    14    15    16    17  \n",
      "A  0.14  0.29  0.26  0.29  0.01  \n",
      "C  0.52  0.03  0.01  0.29  0.04  \n",
      "G  0.18  0.26  0.40  0.42  0.65  \n",
      "U  0.16  0.42  0.33  0.00  0.30  \n"
     ]
    }
   ],
   "source": [
    "#develop scoring matrix\n",
    "\n",
    "\n",
    "#load true and false start spaces\n",
    "trueStartSpace=[]\n",
    "falseStartSpace=[]\n",
    "#first 10 of flat files\n",
    "for i in range(100):\n",
    "    spaces=parseFile(flatfiles[i],path)\n",
    "    trueStartSpace+=spaces[\"trueStarts\"]\n",
    "    falseStartSpace+=spaces[\"falseStarts\"]\n",
    "\n",
    "numNucs=18\n",
    "#build consensus matrices\n",
    "print(\"True start codon identities by position\")\n",
    "startIdentities={\"A\": [0 for i in range(numNucs)], \"U\": [0 for i in range(numNucs)], \"G\": [0 for i in range(numNucs)], \"C\": [0 for i in range(numNucs)]}\n",
    "for i in trueStartSpace:\n",
    "    temp=i.replace(\" \",\"\").upper()\n",
    "    for n in range(numNucs):\n",
    "        try:\n",
    "            startIdentities[temp[n]][n]+=1\n",
    "        except:\n",
    "            print(\"except: \" + temp)\n",
    "#make data frame of start identities for clean format of output\n",
    "startIDsDF=pd.DataFrame.from_dict(startIdentities)\n",
    "print(startIDsDF.T)\n",
    "\n",
    "print(\"\\nTrue start codon frequency by position\")\n",
    "for key in startIdentities:\n",
    "    for i in range(numNucs):\n",
    "        startIdentities[key][i]=round(startIdentities[key][i]/len(trueStartSpace),2)\n",
    "startFreqsDF=pd.DataFrame.from_dict(startIdentities)\n",
    "print(startFreqsDF.T)\n",
    "\n",
    "#print(\"\\nFalse start codon identities by position\")\n",
    "falseStartIdentities={\"A\": [0 for i in range(numNucs)], \"U\": [0 for i in range(numNucs)], \"G\": [0 for i in range(numNucs)], \"C\": [0 for i in range(numNucs)]}\n",
    "for i in falseStartSpace:\n",
    "    temp=i.replace(\" \",\"\").upper()\n",
    "    for n in range(numNucs):\n",
    "        falseStartIdentities[temp[n]][n]+=1\n",
    "#make data frame of false start identities for clean format of output\n",
    "falseStartIDsDF=pd.DataFrame.from_dict(falseStartIdentities)\n",
    "#print(falseStartIDsDF.T)\n",
    "\n",
    "#print(\"\\nFalse start codon frequency by position\")\n",
    "for key in falseStartIdentities:\n",
    "    for i in range(numNucs):\n",
    "        falseStartIdentities[key][i]=round(falseStartIdentities[key][i]/len(falseStartSpace),2)\n",
    "falseStartFreqsDF=pd.DataFrame.from_dict(falseStartIdentities)\n",
    "#print(falseStartFreqsDF.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix=startFreqsDF\n",
    "#score a sequence based on scoreing matrix\n",
    "def scoreSequence(seq, matrix):\n",
    "    seq=seq.upper()\n",
    "    score=1\n",
    "    for i in range(numNucs):\n",
    "        positionScore=matrix[seq[i]][i]\n",
    "        if positionScore==0:\n",
    "            score*=.000001\n",
    "        else:\n",
    "            score*=positionScore\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True start Scores\n",
      "ugaaaugaaauggggucuAF494042.txt\n",
      "ugaaaugaaauggggucuAY150312.txt\n",
      "ugaaaugaaauggggucuAY262352.txt\n",
      "uuuggaguaaugcaauggAY366525.txt\n",
      "aguaaugcaauggggucaAY366525.txt\n",
      "cuucagacaauggcucauAY366525.txt\n",
      "cuuggcgauaugagauguAY366525.txt\n",
      "ggcugagcaaugggaggcAY366525.txt\n",
      "ugaaaugaaaugggguccAY424271.txt\n",
      "ugaaaugaaaugggguccAY585241.txt\n",
      "uucggggugaugcaauggAY588319.txt\n",
      "ggugaugcaauggggucaAY588319.txt\n",
      "uuucagacaauggcucauAY588319.txt\n",
      "cuuggcaauaugagauguAY588319.txt\n",
      "ggcugagcaaugggaggcAY588319.txt\n",
      "ugaaaugaaaugggguccAY612613.txt\n",
      "ugaaaugaaaugggguccDQ473474.txt\n",
      "agcggaacaauggggucaDQ473474.txt\n",
      "cuuggagugaugcaauggDQ489311.txt\n",
      "agugaugcaauggggucaDQ489311.txt\n",
      "cuucagacaauggcucauDQ489311.txt\n",
      "cuuggcgauaugaaauguDQ489311.txt\n",
      "ggccgagcaaugggaggcDQ489311.txt\n",
      "uccgaagugaugcaauggEU076704.txt\n",
      "agugaugcaauggggucaEU076704.txt\n",
      "cuucagacaauggcucguEU076704.txt\n",
      "ccuggcgauaugagauguEU076704.txt\n",
      "ggcugagcaaugggaggcEU076704.txt\n",
      "ugaaaugaaaugggguccEU360128.txt\n",
      "ugaaaugaaaugggguccEU360129.txt\n",
      "ugaaaugaaauggggucuEU360130.txt\n",
      "ugaaaugaaaugggguccEU624117.txt\n",
      "ugaaaugaaauggggucuEU807840.txt\n",
      "agcggaacaaugggcucgEU880431.txt\n",
      "agcggaacaaugggcucgEU880432.txt\n",
      "agcggaacaaugggcucgEU880433.txt\n",
      "agcggaacaaugggcucgEU880434.txt\n",
      "agcggaacaaugggcucgEU880435.txt\n",
      "agcggaacaaugggcucgEU880436.txt\n",
      "agcggaacuaugggcucgEU880437.txt\n",
      "agcggaacaaugggcucgEU880438.txt\n",
      "agcggaacaaugggcucgEU880439.txt\n",
      "agcggaacaaugggcucgEU880440.txt\n",
      "agcggaacaaugggcucgEU880441.txt\n",
      "agcggaacaaugggcucgEU880442.txt\n",
      "agcggaacaaugggcucgEU880443.txt\n",
      "ccugaagugaugcaauggFJ349261.txt\n",
      "agugaugcaauggggucgFJ349261.txt\n",
      "uuucagacaauggcugauFJ349261.txt\n",
      "cuuggcgauaugagauguFJ349261.txt\n",
      "agcugagcaaugggaagcFJ349261.txt\n",
      "ggcggaacaauggggucgFJ394029.txt\n",
      "ugaaaugaaauggggucuGQ351601.txt\n",
      "uccggagugaugcaauggGQ461593.txt\n",
      "uuucagacaauggcucauGQ461593.txt\n",
      "cuuggcgauaugagauguGQ461593.txt\n",
      "ggcugagcaaugggaagcGQ461593.txt\n",
      "uucagagugaugcaauggGU047344.txt\n",
      "agugaugcaauggggucaGU047344.txt\n",
      "cuucaggcaauggcucauGU047344.txt\n",
      "cuuggcgauaugaaauguGU047344.txt\n",
      "agcugagcaaugggaagcGU047344.txt\n",
      "uccgaagugaugcaauggGU047345.txt\n",
      "agugaugcaauggggucaGU047345.txt\n",
      "cuucaggcaauggcucguGU047345.txt\n",
      "cuuggcgauaugagauguGU047345.txt\n",
      "ggcugaacaaugggaagcGU047345.txt\n",
      "uccggagugaugcaauggGU067771.txt\n",
      "agugaugcaaugggguuaGU067771.txt\n",
      "uuucagacaauggcucauGU067771.txt\n",
      "cuuggcgauaugagauguGU067771.txt\n",
      "ggcugagcaaugggaagcGU067771.txt\n",
      "agcggaacgauggggucgGU143913.txt\n",
      "False Start Scores:\n",
      "True high: 4.26915206453488e-06\n",
      "True low: 4.212218543202111e-16\n",
      "False high: 6.743791352490885e-07\n",
      "False Low: 2.755584000000001e-33\n"
     ]
    }
   ],
   "source": [
    "#true start scores\n",
    "print(\"True start Scores\")\n",
    "trueLow=99999999\n",
    "trueHigh=0\n",
    "falseLow=999999\n",
    "falseHigh=0\n",
    "for i in range(len(trueStartSpace)):\n",
    "    s=scoreSequence(trueStartSpace[i],matrix)\n",
    "    if s<trueLow:\n",
    "        trueLow=s\n",
    "    if s>trueHigh:\n",
    "        trueHigh=s\n",
    "    if s<.00000001:\n",
    "        print(trueStartSpace[i])\n",
    "    #print(s)\n",
    "#false start scores\n",
    "print(\"False Start Scores:\")\n",
    "for i in range(len(falseStartSpace)):\n",
    "    s=scoreSequence(falseStartSpace[i],matrix)\n",
    "    if s<falseLow:\n",
    "        falseLow=s\n",
    "    if s>falseHigh:\n",
    "        falseHigh=s\n",
    "    #print(s)\n",
    "\n",
    "print(\"True high: \" + str(trueHigh))\n",
    "print(\"True low: \" + str(trueLow))\n",
    "print(\"False high: \" + str(falseHigh))\n",
    "print(\"False Low: \" + str(falseLow))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
